### PR TITLE
Inline superClassTest for transformed checkCast nodes when castClass is runtime variable

### PR DESCRIPTION
**Pull request by rmnattas**
The case of castClass being a runtime variable was only for `instanceOf` nodes until #15079 (transforms Class.cast calls to checkCast nodes), for this new case we can inline superClassTest to improve performance. This part of common code is used by Z, P and AARCH64. 

This PR allows `checkCast` nodes in the case of variable castClass to inline superClassTest. 
It also adds array and interface checks for castClass before the superClassTest, if so skip superClassTest, for P and AARCH64. It's already implemented in Z.

Description issue: #15535

- ~Finish/check AARCH64 code (using/not-using `tbz/tbnz`, immediate `immr:imms`)~
- ~P handle unknown classDepth for superClassTest~
- ~Testing and benchmark performance~

### CODE REVIEW

**Comment by rmnattas** [comment]
@r30shah @ymanton @0xdaryl

File: runtime/compiler/p/codegen/J9TreeEvaluator.cpp [hunk 1]
```diff
@@ -3549,6 +3549,38 @@ void genInstanceOfOrCheckCastSuperClassTest(TR::Node *node, TR::Register *condRe
...
+         generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::andi_r, node, scratchRegister2, scratchRegister1, condReg, J9AccInterface);
+      else
+         generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::andis_r, node, scratchRegister2, scratchRegister2, condReg, J9AccInterface >> 16);
+      // At this point cr0[eq] will be set if this is not an interface.
+      generateConditionalBranchInstruction(cg, TR::InstOpCode::bne, node, falseLabel, condReg);
+
+      srm->reclaimScratchRegister(scratchRegister1);
+      srm->reclaimScratchRegister(scratchRegister2);
+      }
+
```
Comment (ymanton) [inline runtime/compiler/p/codegen/J9TreeEvaluator.cpp:3583]: If `castClassDepth == -1` is true at compile-time do we have to generate code to load it at run-time? The aarch64 codegen has

```
...

 *  @param[in] castClassDepth:                 class depth of the cast class. If -1 is passed, depth is loaded at runtime

...

   if (castClassDepth != -1)
      {
      // castClassDepth is known at compile time
      ...
      }
   else
      {
      // castClassDepth needs to be loaded from castClass
      castClassDepthReg = srm->findOrCreateScratchRegister();
      // load lower 16bit of classDepthAndFlags
      generateTrg1MemInstruction(cg, TR::InstOpCode::ldrhimm, node, castClassDepthReg,
                              TR::MemoryReference::createWithDisplacement(cg, castClassReg, offsetof(J9Class, classDepthAndFlags)));
      generateCompareInstruction(cg, node, instanceClassDepthReg, castClassDepthReg);
      }
```

which loads `castClassReg->classDepthAndFlags` but the Power codegen only has

```
   if (castClassDepth <= UPPER_IMMED && castClassDepth >= LOWER_IMMED)
      generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::cmpi4, node, condReg, instanceClassDepthReg, castClassDepth);
   else
      {
      castClassDepthReg = srm->findOrCreateScratchRegister();
      loadConstant(cg, node, castClassDepth, castClassDepthReg);
      generateTrg1Src2Instruction(cg, TR::InstOpCode::cmp4, node, condReg, instanceClassDepthReg, castClassDepthReg);
      }
```

which is generating code to use `castClassDepth` either as an immediate if it fits or via a register otherwise, but not doing anything based on whether it is `-1` or not.

Comment (rmnattas) [reply runtime/compiler/p/codegen/J9TreeEvaluator.cpp:3583]: Correct, good catch. The current superClassTest inlining code does not handle unknown classDepth and generates useless code. Adding to my WIP list.

Comment (rmnattas) [reply runtime/compiler/p/codegen/J9TreeEvaluator.cpp:3583]: Fixed in 8000695e619da88a7ecb84ca27cd7e3524988fac

**Comment by rmnattas** [comment]
@knn-k Can you have a look at the AArch64 code, espically using/not-using `tbz/tbnz` and advice on the `immr:imms` immediate format. I'm confused about the immediate as I found a way for it but it doesn't confirm with the existing `0x400 is immr:imms for 0x10000`.

File: runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp [hunk 2]
```diff
@@ -1746,6 +1746,39 @@ static
...
+      if (node->->getOpCodeValue() == TR::instanceof)
+         {
+         generateTestBitBranchInstruction(cg, TR::InstOpCode::tbnz, node, scratchRegister1, 16, falseLabel);
+         generateTestBitBranchInstruction(cg, TR::InstOpCode::tbnz, node, scratchRegister1, 9, falseLabel);
+         }
+      else
+         {
+         generateTestImmInstruction(cg, node, scratchRegister1, 0x400); // 0x400 is immr:imms for 0x10000
+         generateConditionalBranchInstruction(cg, TR::InstOpCode::b_cond, node, falseLabel, TR::CC_NE);
+         generateTestImmInstruction(cg, node, scratchRegister1, 0xDEADBEEF); // 0xXXXX is immr:imms for 0x200
```
Comment (knn-k) [inline runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp:1775]: Please use 0x5c0 as `immr:imms` for `tst w0, #0x200`.

File: runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp [hunk 3]
```diff
@@ -1746,6 +1746,39 @@ static
...
+            "genSuperClassTest: castClassDepth == -1 is not supported for a loadaddr castClass");
+
+      generateTrg1MemInstruction(cg, TR::InstOpCode::ldrimmx, node, scratchRegister1, TR::MemoryReference::createWithDisplacement(cg, castClassReg, offsetof(J9Class, romClass)));
+      generateTrg1MemInstruction(cg, TR::InstOpCode::ldrimmw, node, scratchRegister1, TR::MemoryReference::createWithDisplacement(cg, scratchRegister1, offsetof(J9ROMClass, modifiers)));
+
+      // Array an check
+      static_assert(J9AccClassArray == 0x10000, "J9AccClassArray must be 0x10000");
+      static_assert(J9AccInterface == 0x200, "J9AccInterface must be 0x200");
+      // Two cases can reach this code, an instanceOf node or a call to Class.cast() converted to a checkCast node
+      if (node->->getOpCodeValue() == TR::instanceof)
```
Comment (knn-k) [inline runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp:1766]: Do you need this condition?
Instructions `tbnz reg, 16, falseLabel` and `tst reg, #0x10000 ; b.ne falseLabel` mean the same, except for the status of condition flags.

File: runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp [hunk 4]
```diff
@@ -1746,6 +1746,39 @@ static
...
+   if ( dynamicCastClass )
+      {
+      TR::Register *scratchRegister1 = srm->findOrCreateScratchRegister();
+      TR_ASSERT(node->getSecondChild()->getOpCodeValue() != TR::loadaddr,
+            "genSuperClassTest: castClassDepth == -1 is not supported for a loadaddr castClass");
+
+      generateTrg1MemInstruction(cg, TR::InstOpCode::ldrimmx, node, scratchRegister1, TR::MemoryReference::createWithDisplacement(cg, castClassReg, offsetof(J9Class, romClass)));
+      generateTrg1MemInstruction(cg, TR::InstOpCode::ldrimmw, node, scratchRegister1, TR::MemoryReference::createWithDisplacement(cg, scratchRegister1, offsetof(J9ROMClass, modifiers)));
+
+      // Array an check
```
Comment (knn-k) [inline runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp:1762]: I don't think you need "an" here.

File: runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp (hunk 3 above)
Comment (rmnattas) [reply runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp:1766]: I noticed in the code that instanceOf uses `tbz/tbnz` and checkCast uses `tst; b.ne`, I believe as the target for `tbz/tbnz` must be within 32KB and that doesn't work for checkCast, and tried to keep it the same. 
I guess I should ask if there's a benefit for `tbz/tbnz` to do it this way, or just use `tst; b.ne` for both cases here?

Comment (knn-k) [reply runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp:1766]: `tbnz` is preferred when the destination is within the ±32KB range, because it is faster and it consumes less memory than `tst; b.ne` for this case.
Is the destination `falseLabel` far away for the checkCast case?

Comment (knn-k) [reply runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp:1766]: Here is an example.  `genInstanceOfOrCheckCastObjectArrayTest()` takes a parameter `useTBZ` for deciding which code to use.

https://github.com/eclipse-openj9/openj9/blob/ae6f910e02511bd082b55b821036c435e00517e1/runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp#L1891-L1899

Comment (rmnattas) [reply runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp:1766]: It looks like even in the instanceOf case, the `falseLabel` can be a OOLCodeSection and can be placed far away. So we should use `tst; b.ne` for all cases.

File: runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp (hunk 4 above)
Comment (rmnattas) [reply runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp]: Fixed

File: runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp (hunk 3 above)
Comment (rmnattas) [reply runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp]: Corrected the code in b0255bc69f698fa27460f100006637d5808a45c9

File: runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp (hunk 2 above)
Comment (rmnattas) [reply runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp]: Great, thank you

**Comment by rmnattas** [comment]
@ymanton Please review the latest P code.

P testing [HERE](https://hyc-runtimes-jenkins.swg-devops.com/job/Pipeline-Build-Test-Personal/13947/) (the 2 non-passing extended tests are not caused by this change)
Will run Z and AArch64 next

**Comment by knn-k** [comment]
I thought the check code could be simpler as:
```
if ((modifiers & (J9AccClassArray | J9AccInterface)) != 0)
	goto falseLabel;
```

But it does not work for AArch64, because you cannot embed `0x10200` into the `tst (immediate)` instruction.

**Comment by rmnattas** [comment]
Updated the P code for a faster sequence:
```
Original Sequence:
 [0xa00000040210830]    11      andis.  GPR_0083, GPR_0083, 1
 [0xa00000040210930]    11      bne     CCR_0080, Outlined Label L0060
 [0xa000000402109d0]    11      andi.   GPR_0083, GPR_0082, 512
 [0xa00000040210ad0]    11      bne     CCR_0080, Outlined Label L0060

New Sequence:
 [    0x7fff65f0dc00]   11      lis     GPR_0068, 0000000000000001
 [    0x7fff65f0dca0]   11      ori     GPR_0068, GPR_0068, 512
 [    0x7fff65f0dd40]   11      and.    GPR_0068, GPR_0067, GPR_0068
 [    0x7fff65f0dde0]   11      bne     CCR_0065, Outlined Label L0044
```

**Comment by rmnattas** [comment]
@knn-k Is a similar sequence to the last comment possible/beneficial in AArch64?
```
load 0x10200 into a register
and with compare flags set
branch
```

**Comment by rmnattas** [comment]
**For benchmark and performance**

Running a microbenchmark for the execution time of the `Class.cast(obj)` call (measured on P):
There's 28% improvement when the castClass is a superClass of the object (for the superClassTest inlined sequence)
There's 3-4% performance hit when the castClass is not equal or a superClass of the object (for the inlined checks)

For the inlined superClassTest fix for `instanceOf` for P:
There's 20% improvement when the castClass is a superClass of the object
There's no performance hit as the current inlined code exist but ineffective

**Comment by knn-k** [comment]
> Is a similar sequence to the last comment possible/beneficial in AArch64?

```
loadConstant32(cg, node, (J9AccClassArray | J9AccInterface), scratchRegister2); // This requires two instructions on AArch64
generateTestInstruction(cg, node, scratchRegister1, scratchRegister2); // 32-bit comparison
generateConditionalBranchInstruction(cg, TR::InstOpCode::b_cond, node, falseLabel, TR::CC_NE);
```

You need only one conditional branch with this sequence, while you need an extra temporary register.

**Comment by rmnattas** [comment]
Updated AArch64 sequence
```
Original Sequence
[    0xffff80ddd4c0] 7210007f 11    tstimmw         w3, 0x10000
[    0xffff80ddd550] 54000ba1 11    b.ne    Outlined Label L0038    
[    0xffff80ddd5e0] 7217007f 11    tstimmw         w3, 0x200
[    0xffff80ddd670] 54000b61 11    b.ne    Outlined Label L0038

New sequence
[    0xffff8e8d3550] 52804002 11    movzw   w2, 0x0200
[    0xffff8e8d35e0] 72a00022 11    movkw   w2, 0x0001, LSL #16
[    0xffff8e8d3670] 6a02007f 11    tstw    w3, w2
[    0xffff8e8d3700] 54000b61 11    b.ne    Outlined Label L0038
```

**Comment by rmnattas** [comment]
This PR is ready to merge

File: runtime/compiler/codegen/J9TreeEvaluator.cpp [hunk 5]
```diff
@@ -1346,7 +1346,6 @@ uint32_t J9::TreeEvaluator::calculateInstanceOfOrCheckCastSequences(TR::Node *in
       sequences[i++] = CastClassCacheTest;
       // On Z, We were having support for Super Class Test for dynamic Cast Class so adding it here. It can be guarded if Power/X do not need it.
       if (cg->supportsInliningOfIsInstance() &&
```
[the remark this answers is not in the document]
Comment (rmnattas) [reply runtime/compiler/codegen/J9TreeEvaluator.cpp]: Hmm... I guess so. I would have thought it'll apply to `checkCast` too given both nodes inline roughly the same thing. It doesn't seem like there's a similar option for `checkCast`. Either way, if that's what the option suppose to do makes sense.

File: runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp [hunk 6]
```diff
@@ -1746,6 +1746,31 @@ static
...
+      TR::Register *scratchRegister2 = srm->findOrCreateScratchRegister();
+      TR_ASSERT(node->getSecondChild()->getOpCodeValue() != TR::loadaddr,
+            "genSuperClassTest: castClassDepth == -1 is not supported for a loadaddr castClass");
+
+      generateTrg1MemInstruction(cg, TR::InstOpCode::ldrimmx, node, scratchRegister1, TR::MemoryReference::createWithDisplacement(cg, castClassReg, offsetof(J9Class, romClass)));
+      generateTrg1MemInstruction(cg, TR::InstOpCode::ldrimmw, node, scratchRegister1, TR::MemoryReference::createWithDisplacement(cg, scratchRegister1, offsetof(J9ROMClass, modifiers)));
+
+      // Array and interface check
+      static_assert(J9AccClassArray == 0x10000, "J9AccClassArray must be 0x10000");
+      static_assert(J9AccInterface == 0x200, "J9AccInterface must be 0x200");
```
Comment (knn-k) [inline runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp:1765]: I don't think you need these two `static_assert()` lines when you load the value `(J9AccClassArray | J9AccInterface)` in a register using `loadConstant32()`.

Comment (rmnattas) [reply runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp]: Removed the exact value asserts but added an assert to check that `(J9AccInterface | J9AccClassArray)` is a 32-bit value.

**Comment by knn-k** [comment]
The AArch64 code looks good to me.

**Review by ymanton — APPROVED** [review]
Power changes LGTM.

**Comment by ymanton** [comment]
Jenkins test sanity.functional all jdk8,jdk11

**Comment by ymanton** [comment]
Tests failed due to infrastructure issues, for example:

```
17:29:24  Cannot contact cent6-x64-1: java.lang.InterruptedException
```

I'll retry tomorrow.

**Comment by knn-k** [comment]
Jenkins test sanity.functional all jdk8,jdk11

**Comment by knn-k** [comment]
https://openj9-jenkins.osuosl.org/job/Test_openjdk11_j9_sanity.functional_s390x_linux_Personal_testList_0/206/
https://openj9-jenkins.osuosl.org/job/Test_openjdk11_j9_sanity.functional_s390x_linux_Personal_testList_1/198/
https://openj9-jenkins.osuosl.org/job/Test_openjdk8_j9_sanity.functional_s390x_linux_Personal_testList_0/176/
https://openj9-jenkins.osuosl.org/job/Test_openjdk8_j9_sanity.functional_s390x_linux_Personal_testList_1/176/

Jobs on s390x_linux above failed with the following error in compiling `org/apache/tools/ant/types/Resource.as(Ljava/lang/Class;)Ljava/lang/Object;`.  I don't know if they are caused by this PR or not.

```
[2022-08-26T06:01:48.798Z] Unhandled exception
[2022-08-26T06:01:48.798Z] Type=Segmentation error vmState=0x0005ff06
[2022-08-26T06:01:48.798Z] J9Generic_Signal_Number=00000018 Signal_Number=0000000b Error_Value=f0b33e48 Signal_Code=00000001
[2022-08-26T06:01:48.798Z] Handler1=000003FF9DA47B28 Handler2=000003FF9D932F78 InaccessibleAddress=0000000000000000
[2022-08-26T06:01:48.798Z] gpr0=0000000000000000 gpr1=000003FF9CF97470 gpr2=0000000000000000 gpr3=000003FF9D2F23F0
[2022-08-26T06:01:48.798Z] gpr4=000003FF9CFB58E4 gpr5=0000000000000000 gpr6=0000000000000000 gpr7=0000000000000000
[2022-08-26T06:01:48.798Z] gpr8=0000000000000000 gpr9=000003FF9D339628 gpr10=000003FF9D336B00 gpr11=000003FF0D0D8890
[2022-08-26T06:01:48.798Z] gpr12=000003FF9D335000 gpr13=0000000000000000 gpr14=000003FF9CFB58E4 gpr15=000003FF756760E8
[2022-08-26T06:01:48.798Z] psw=000003FF9CFB595A mask=0705000180000000 fpc=0088fe00 bea=000003FF9CFB58FE
[2022-08-26T06:01:48.798Z] fpr0 409e000000000000 (f: 0.000000, d: 1.920000e+03)
[2022-08-26T06:01:48.798Z] fpr1 3fb5f15f15f15f16 (f: 368140064.000000, d: 8.571429e-02)
[2022-08-26T06:01:48.798Z] fpr2 0000000000000000 (f: 0.000000, d: 0.000000e+00)
[2022-08-26T06:01:48.798Z] fpr3 3fd999cbe9982871 (f: 3919063040.000000, d: 4.000120e-01)
[2022-08-26T06:01:48.798Z] fpr4 409e000000000000 (f: 0.000000, d: 1.920000e+03)
[2022-08-26T06:01:48.798Z] fpr5 3fcc720a74b7854b (f: 1958184320.000000, d: 2.222303e-01)
[2022-08-26T06:01:48.798Z] fpr6 3fc75ecaddb24359 (f: 3719447296.000000, d: 1.825803e-01)
[2022-08-26T06:01:48.798Z] fpr7 3f7e17d2dc43b59b (f: 3695425024.000000, d: 7.346939e-03)
[2022-08-26T06:01:48.798Z] fpr8 00000000102e5be8 (f: 271473632.000000, d: 1.341258e-315)
[2022-08-26T06:01:48.798Z] fpr9 0000000000000000 (f: 0.000000, d: 0.000000e+00)
[2022-08-26T06:01:48.798Z] fpr10 00000000102e5250 (f: 271471168.000000, d: 1.341246e-315)
[2022-08-26T06:01:48.798Z] fpr11 0000000000000000 (f: 0.000000, d: 0.000000e+00)
[2022-08-26T06:01:48.798Z] fpr12 00000000102e5bf0 (f: 271473664.000000, d: 1.341258e-315)
[2022-08-26T06:01:48.798Z] fpr13 0000000000000000 (f: 0.000000, d: 0.000000e+00)
[2022-08-26T06:01:48.798Z] fpr14 0000000000000000 (f: 0.000000, d: 0.000000e+00)
[2022-08-26T06:01:48.798Z] fpr15 0000000000000000 (f: 0.000000, d: 0.000000e+00)
[2022-08-26T06:01:48.798Z] Module=/home/jenkins/workspace/Test_openjdk11_j9_sanity.functional_s390x_linux_Personal_testList_0/openjdkbinary/j2sdk-image/lib/default/libj9jit29.so
[2022-08-26T06:01:48.798Z] Module_base_address=000003FF9C500000
[2022-08-26T06:01:48.798Z] 
[2022-08-26T06:01:48.798Z] Method_being_compiled=org/apache/tools/ant/types/Resource.as(Ljava/lang/Class;)Ljava/lang/Object;
[2022-08-26T06:01:48.798Z] Target=2_90_20220825_494 (Linux 3.10.0-1160.71.1.el7.s390x)
[2022-08-26T06:01:48.798Z] CPU=s390x (4 logical CPUs) (0x1ec1b4000 RAM)
[2022-08-26T06:01:48.798Z] ----------- Stack Backtrace -----------
[2022-08-26T06:01:48.798Z] _ZN2TR21S390BranchInstruction30assignRegistersAndDependenciesE16TR_RegisterKinds+0x1d2 (0x000003FF9CFB595A [libj9jit29.so+0xab595a])
[2022-08-26T06:01:48.798Z] _ZN3OMR1Z11Instruction15assignRegistersE16TR_RegisterKinds+0x2c (0x000003FF9D01B1D4 [libj9jit29.so+0xb1b1d4])
[2022-08-26T06:01:48.798Z] _ZN3OMR13CodeGenerator20doRegisterAssignmentE16TR_RegisterKinds+0xec (0x000003FF9CB16D7C [libj9jit29.so+0x616d7c])
[2022-08-26T06:01:48.798Z] _ZN3OMR12CodeGenPhase29performRegisterAssigningPhaseEPN2TR13CodeGeneratorEPNS1_12CodeGenPhaseE+0xa4 (0x000003FF9CB1D094 [libj9jit29.so+0x61d094])
[2022-08-26T06:01:48.798Z] _ZN3OMR12CodeGenPhase10performAllEv+0x10e (0x000003FF9CB1E1CE [libj9jit29.so+0x61e1ce])
[2022-08-26T06:01:48.798Z] _ZN3OMR13CodeGenerator12generateCodeEv+0x60 (0x000003FF9CB1A828 [libj9jit29.so+0x61a828])
[2022-08-26T06:01:48.798Z] _ZN3OMR11Compilation7compileEv+0xe28 (0x000003FF9CB4D888 [libj9jit29.so+0x64d888])
[2022-08-26T06:01:48.799Z] _ZN2TR28CompilationInfoPerThreadBase7compileEP10J9VMThreadPNS_11CompilationEP17TR_ResolvedMethodR11TR_J9VMBaseP19TR_OptimizationPlanRKNS_16SegmentAllocatorE+0x4aa (0x000003FF9C686622 [libj9jit29.so+0x186622])
[2022-08-26T06:01:48.799Z] _ZN2TR28CompilationInfoPerThreadBase14wrappedCompileEP13J9PortLibraryPv+0x35a (0x000003FF9C6876BA [libj9jit29.so+0x1876ba])
[2022-08-26T06:01:48.799Z] omrsig_protect+0x366 (0x000003FF9D9340E6 [libj9prt29.so+0x340e6])
[2022-08-26T06:01:48.799Z] _ZN2TR28CompilationInfoPerThreadBase7compileEP10J9VMThreadP21TR_MethodToBeCompiledRN2J917J9SegmentProviderE+0x332 (0x000003FF9C684D0A [libj9jit29.so+0x184d0a])
[2022-08-26T06:01:48.799Z] _ZN2TR24CompilationInfoPerThread12processEntryER21TR_MethodToBeCompiledRN2J917J9SegmentProviderE+0x1d2 (0x000003FF9C6853B2 [libj9jit29.so+0x1853b2])
[2022-08-26T06:01:48.799Z] _ZN2TR24CompilationInfoPerThread14processEntriesEv+0x3c6 (0x000003FF9C683DCE [libj9jit29.so+0x183dce])
[2022-08-26T06:01:48.799Z] _ZN2TR24CompilationInfoPerThread3runEv+0x66 (0x000003FF9C6842AE [libj9jit29.so+0x1842ae])
[2022-08-26T06:01:48.799Z] _Z30protectedCompilationThreadProcP13J9PortLibraryPN2TR24CompilationInfoPerThreadE+0x9a (0x000003FF9C684392 [libj9jit29.so+0x184392])
[2022-08-26T06:01:48.799Z] omrsig_protect+0x366 (0x000003FF9D9340E6 [libj9prt29.so+0x340e6])
[2022-08-26T06:01:48.799Z] _Z21compilationThreadProcPv+0x20a (0x000003FF9C68483A [libj9jit29.so+0x18483a])
[2022-08-26T06:01:48.799Z] thread_wrapper+0x114 (0x000003FF9D885A14 [libj9thr29.so+0x5a14])
[2022-08-26T06:01:48.799Z] start_thread+0xea (0x000003FF9E388312 [libpthread.so.0+0x8312])
[2022-08-26T06:01:48.799Z]  (0x000003FF9E18E232 [libc.so.6+0x10e232])
[2022-08-26T06:01:48.799Z] ---------------------------------------
```

**Comment by r30shah** [comment]
I have a strong suspicion that failures on Z are related to this change. Unfortunately I can not tell more as the failure happens while building test and none of them uploads any result files. Let me cherry pick this change and launch a build internally to see if it is reproducible

**Comment by r30shah** [comment]
Failure on Z was 100% reproducible. I looked through the log file and found that when we emit super class test on Z for checkCast node with dynamic cast class, we emitted a branch instruction with NULL call label (Instead of helper call label). I have fixed that in https://github.com/r30shah/openj9/commit/0623889dd8503703ae67f80b29764450faa6ebd5. 
@rmnattas  Do you want to include this change as part of this PR?

**Comment by rmnattas** [comment]
@r30shah Added the commit to this PR, and rebased it before my changes.
@ymanton Can you run the tests again.

**Comment by ymanton** [comment]
Jenkins test sanity.functional all jdk8,jdk11

**Comment by ymanton** [comment]
Jenkins test sanity.functional all jdk8,jdk11

**Comment by ymanton** [comment]
@rmnattas can you open a PR for the `v0.35.0-release` branch as well?
